File: quick_sort.cpp
#include <iostream>
using namespace std;

void quick_sort(int *a, int l, int r)
{
    if (l >= r)
        return;
    int i = l;
    int j = r + 1;
    //开头的数为基准数
    int key = a[l];
    while (true)
    {
        // 从左向右找比key大的值
        while (a[++i] < key) 
            if (i == r)
                break;
        // 从右向左找比key小的值
        while (a[--j] > key)
            if (j == l)
                break;
        if (i >= j)
            break;
        int t = a[i];
        a[i] = a[j];
        a[j] = t;
    }
    // 基准数与j对应值交换
    int t = a[l];
    a[l] = a[j];
    a[j] = t;
    quick_sort(a, l, j - 1);
    quick_sort(a, j + 1, r);
}
int main()
{
    int a[9] = {1, -1, 3, 5, 7, 8, 4, 6, 2};
    quick_sort(a, 0, 8);
    for (int i = 0; i < 9; i++)
        cout << a[i] << " ";
    system("pause");
    return 0;
}

File: quick_sort.js
let arr = [1, 3, 5, 4, 9, 6];

function quick_sort(arr) {
  // 边界检查
  if (arr.length <= 1) {
    return arr;
  }
  // 选取基准
  let pivotIndex = Math.floor(arr.length / 2);
  let pivot = arr.splice(pivotIndex, 1)[0];
  // 划分子集
  let left = [];
  let right = [];
  for (let i = 0; i < arr.length; i++) {
    if (arr[i] < pivot) {
      left.push(arr[i]);
    }else{
      right.push(arr[i]);
    }
  }
  // 递归进行
  return quick_sort(left).concat([pivot], quick_sort(right));
}

console.log(quick_sort(arr));